File: SciChart.Examples/Examples/Examples2D/TooltipsAndHitTest/ChangeColumnColorChartView/ChangeColumnColorChartView.m
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2023. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// ChangeColumnColorChartView.m is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

#import "ChangeColumnColorChartView.h"
#import "CustomSeriesInfoProvider.h"

#pragma mark - Color Palette Provider
@interface ColumnsPaletteProvider : SCIPaletteProviderBase<SCIFastColumnRenderableSeries *><ISCIFillPaletteProvider>
- (instancetype)init;
@property (nonatomic) NSInteger touchedIndex;
@end

@implementation ColumnsPaletteProvider {
    SCIUnsignedIntegerValues *_colors;
}

static unsigned int desiredColors[] = {0xFF21a0d8, 0xFFc43360};

- (instancetype)init {
    self = [super initWithRenderableSeriesType:SCIFastColumnRenderableSeries.class];
    if (self) {
        _colors = [SCIUnsignedIntegerValues new];
        _touchedIndex = -1;
    }
    return self;
}

- (void)update {
    NSInteger count = self.renderableSeries.currentRenderPassData.pointsCount;
    _colors.count = count;
    
    unsigned int *colorsArray = _colors.itemsArray;
    for (NSInteger i = 0; i < count; i++) {
        
        if (i == _touchedIndex) {
            colorsArray[i] = desiredColors[1];
        }
        else {
            colorsArray[i] = desiredColors[0];
        }
    }
}

- (SCIUnsignedIntegerValues *)fillColors {
    return _colors;
}

@end

#pragma mark - Chart Initialisation
@implementation ChangeColumnColorChartView {
    SCIHitTestInfo *_hitTestInfo;
}

- (Class)associatedType { return SCIChartSurface.class; }

- (void)initExample {
    id<ISCIAxis> xAxis = [SCINumericAxis new];
    xAxis.growBy = [[SCIDoubleRange alloc] initWithMin:0.1 max:0.1];
    
    id<ISCIAxis> yAxis = [SCINumericAxis new];
    yAxis.axisAlignment = SCIAxisAlignment_Left;
    yAxis.growBy = [[SCIDoubleRange alloc] initWithMin:0.0 max:0.1];
    
    SCIXyDataSeries *dataSeries = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Double yType:SCIDataType_Double];
    int yValues[] = {50, 35, 61, 58, 50, 50, 40, 53, 55, 23, 45, 12, 59, 60};
    for (size_t i = 0; i < sizeof(yValues)/sizeof(yValues[0]); i++) {
        [dataSeries appendX:@(i) y:@(yValues[i])];
    }
    
    SCIFastColumnRenderableSeries *rSeries = [SCIFastColumnRenderableSeries new];
    rSeries.dataSeries = dataSeries;
    rSeries.dataPointWidth = 0.7;
    rSeries.paletteProvider = [ColumnsPaletteProvider new];
    CustomSeriesInfoProvider *customSeriesInfoProvider = [CustomSeriesInfoProvider new];
    customSeriesInfoProvider.delegate = self;
    rSeries.seriesInfoProvider = customSeriesInfoProvider;
    
    [SCIUpdateSuspender usingWithSuspendable:self.surface withBlock:^{
        [self.surface.xAxes add:xAxis];
        [self.surface.yAxes add:yAxis];
        [self.surface.renderableSeries add:rSeries];
        [self.surface.chartModifiers add:[SCITooltipModifier new]];
    }];
}

- (void)handleSingleTap:(SCITapGestureRecognizer *)recognizer {
    CGPoint location = [recognizer locationInView:recognizer.view];
    
    CGPoint hitTestPoint = [self.surface translatePoint:location hitTestable:self.surface.renderableSeriesArea];

    SCIRenderableSeriesCollection *seriesCollection = self.surface.renderableSeries;
    for (NSInteger i = 0, count = seriesCollection.count; i < count; i++) {
        id<ISCIRenderableSeries> rSeries = seriesCollection[i];
        [rSeries hitTest:_hitTestInfo at:hitTestPoint];

        ColumnsPaletteProvider *paletteProvider = (ColumnsPaletteProvider*)[rSeries paletteProvider];
        
        if (paletteProvider != nil) {
            if (_hitTestInfo.isHit) {
                paletteProvider.touchedIndex = _hitTestInfo.dataSeriesIndex;
            }
            else {
                paletteProvider.touchedIndex = -1;
            }
        }
    }
    [self.surface invalidateElement];
}

-(void)getTouchDataSeriesIndex: (NSInteger)dataSeriesIndex {
    NSLog(@"Index =>%ld", (long)self.surface.renderableSeries);
    
    SCIRenderableSeriesCollection  *seriesCollection = self.surface.renderableSeries;
    ColumnsPaletteProvider *paletteProvider = (ColumnsPaletteProvider*)[[seriesCollection firstObject] paletteProvider];
    
    if (paletteProvider != nil) {
        paletteProvider.touchedIndex = dataSeriesIndex;
        
    }
    
    [self.surface invalidateElement];
}
@end

File: SciChart.Examples/Examples/Examples2D/TooltipsAndHitTest/ChangeColumnColorChartView/ChangeColumnColorChartView.swift
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2023. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// ChangeColumnColorChartView.swift is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

//MARK: - Color Palette Provider
class ColumnsPaletteProvider : SCIPaletteProviderBase<SCIFastColumnRenderableSeries>, ISCIFillPaletteProvider {
    
    let colors = SCIUnsignedIntegerValues()
    let desiredColors:[UInt32] = [0xFF21a0d8, 0xFFc43360]
    var touchedIndex: Int = -1
    
    init() {
        super.init(renderableSeriesType: SCIFastColumnRenderableSeries.self)
    }
    
    override func update() {
        let count = renderableSeries!.currentRenderPassData.pointsCount
        colors.count = count
        
        for i in 0 ..< count {
            if i == touchedIndex {
                colors.set(desiredColors[1], at: i)
            }
            else {
                colors.set(desiredColors[0], at: i)
            }
        }
    }
    
    var fillColors: SCIUnsignedIntegerValues { return colors }
}

//MARK: - Chart Initialisation
class ChangeColumnColorChartView: SCDSingleChartViewController<SCIChartSurface> {
    
    override var associatedType: AnyClass { return SCIChartSurface.self }
    
    private let hitTestInfo = SCIHitTestInfo()
    
    override func initExample() {
        let xAxis = SCINumericAxis()
        xAxis.growBy = SCIDoubleRange(min: 0.1, max: 0.1)
        let yAxis = SCINumericAxis()
        yAxis.axisAlignment = .left
        yAxis.growBy = SCIDoubleRange(min: 0, max: 0.1)
        
        let dataSeries = SCIXyDataSeries(xType: .double, yType: .double)
        let yValues = [50, 35, 61, 58, 50, 50, 40, 53, 55, 23, 45, 12, 59, 60];
        for i in 0 ..< yValues.count {
            dataSeries.append(x: i, y: yValues[i])
        }
        
        let rSeries = SCIFastColumnRenderableSeries()
        rSeries.dataSeries = dataSeries
        rSeries.dataPointWidth = 0.7
        rSeries.paletteProvider = ColumnsPaletteProvider()
        let customSeriesInfoProvider = CustomSeriesInfoProvider()
        customSeriesInfoProvider.delegate = self
        rSeries.seriesInfoProvider = customSeriesInfoProvider

        SCIUpdateSuspender.usingWith(surface) {
            self.surface.xAxes.add(xAxis)
            self.surface.yAxes.add(yAxis)
            self.surface.renderableSeries.add(rSeries)
            self.surface.chartModifiers.add(items: SCITooltipModifier())
        }
    }
}

extension ChangeColumnColorChartView: CustomSeriesInfoProviderDelegate {
    func getTouchDataSeriesIndex(dataSeriesIndex: Int) {
        print("Did touch => \(dataSeriesIndex)")
        let seriesCollection = surface.renderableSeries
        let rSeries = seriesCollection.firstObject
        
            if let paletteProvider = rSeries.paletteProvider as? ColumnsPaletteProvider {
                paletteProvider.touchedIndex =  dataSeriesIndex
            }
        surface.invalidateElement()
    }
}
